bus-mapping/src/circuit_input_builder/input_state_ref.rs

```diff
@@ -1655,37 +1655,53 @@ impl<'a> CircuitInputStateRef<'a> {
         dst_addr: impl Into<MemoryAddress>,
         src_addr_end: impl Into<MemoryAddress>,
         bytes_left: impl Into<MemoryAddress>,
-        memory_updated: &Memory,
     ) -> Result<(CopyEventSteps, CopyEventPrevBytes), Error> {
-        let src_addr = src_addr.into();
-        let dst_addr = dst_addr.into();
-        let src_addr_end = src_addr_end.into();
-        let bytes_left = bytes_left.into();
+        let src_addr = src_addr.into().0;
+        let dst_addr = dst_addr.into().0;
+        let src_addr_end = src_addr_end.into().0;
+        let bytes_left = bytes_left.into().0;
+        let src_copy_end = src_addr + bytes_left;
+        let dst_copy_end = dst_addr + bytes_left;
 
         if bytes_left == 0 {
             return Ok((vec![], vec![]));
         }
 
+        // Extend call memory.
+        let memory = &mut self.call_ctx_mut()?.memory;
+        memory.extend_for_range(dst_addr.into(), bytes_left.into());
+
+        // Align memory range.
         let dst_range = MemoryWordRange::align_range(dst_addr, bytes_left);
-        let code_slot_bytes = memory_updated.read_chunk(dst_range);
+        let dst_begin_slot = dst_range.start_slot().0;
+        let dst_end_slot = dst_range.end_slot().0.min(memory.len());
+        assert!(dst_begin_slot <= dst_addr && dst_end_slot >= dst_copy_end);
```

Comment:
memory.len cannot be less than end_slot after extend. If it is that is a bug; instead of taking min().

Reply:
Replace memory `extend_for_range(dst_addr, bytes_left)` with `extend_for_range(begin_slot, full_length)` in commit [9b70103](https://github.com/scroll-tech/zkevm-circuits/pull/629/commits/9b70103d77c1ea7c22d6ea2e48768368417c78b8).

Reply:
Ho no that’s not I meant.

`memory.extend_for_range(dst_addr.into(), bytes_left.into());`

This removed variant was the most correct, because it implements pure EVM semantics of memory expansion, while "slot" and "full_length" are our own internal invention. Although this happens to give the same result.

Reply:
I see. Reverted and extract it to a common function [combine_copy_slot_bytes](https://github.com/scroll-tech/zkevm-circuits/blob/7763eddec4cd3083144263bf40a99fd080fa22c3/bus-mapping/src/circuit_input_builder/input_state_ref.rs#L2062) (for these opcodes).